sound/soc/sof/sof-audio.c

```diff
@@ -348,15 +348,16 @@ static int sof_free_widgets_in_path(struct snd_sof_dev *sdev, struct snd_soc_dap
 	int err;
 	int ret = 0;
 
+	if (WIDGET_IS_AIF_OR_DAI(widget->id)) {
+		err = sof_widget_free(sdev, widget->dobj.private);
+		if (err < 0)
+			ret = err;
+	}
+
 	/* free all widgets even in case of error to keep use counts balanced */
 	snd_soc_dapm_widget_for_each_sink_path(widget, p) {
 		if (!p->walking && p->sink->dobj.private && widget->dobj.private) {
 			p->walking = true;
-			if (WIDGET_IS_AIF_OR_DAI(widget->id)) {
-				err = sof_widget_free(sdev, widget->dobj.private);
-				if (err < 0)
-					ret = err;
-			}
 
 			err = sof_widget_free(sdev, p->sink->dobj.private);
```

Comment:
Can we call `sof_widget_free(sdev, widget->dobj.private);` unconditionally outside the loop and remove `sof_widget_free(sdev, p->sink->dobj.private);` here?

Reply:
yes, i've fixed both setup and free functions to unconditionally set up the starting widget outside the loop now